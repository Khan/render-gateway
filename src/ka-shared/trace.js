// @flow
import * as traceAgent from "@google-cloud/trace-agent";
import type {$Request} from "express";
import {trace as traceImpl} from "../shared/index.js";
import {getLogger} from "./get-logger.js";

import type {RequestWithLog, ITraceSession} from "../shared/types.js";

/**
 * Start tracing an event.
 *
 * This will log the start of a trace and open a trace session, which is
 * returned. Use the returned session to end the trace when the traced event is
 * over. The traced event will be logged and also written to the Google Cloud
 * StackDriver Trace agent.
 *
 * Note that if startTraceAgent was never called, this will still log but the
 * StackDriver trace span creation will not actually happen.
 *
 * @param {string} name The name of the event being traced.
 * @param {TReq: RequestWithLog<$Request>} [request] The request being
 * fulfilled. This is used to determine if a request-scoped logger can be used.
 * @returns {ITraceSession} The new trace session that was created and is to be
 * used to end the session.
 */
export const trace = <TReq: RequestWithLog<$Request>>(
    name: string,
    request?: TReq,
): ITraceSession => {
    const logger = getLogger(request);
    const tracer = traceAgent.get();
    return traceImpl(logger, name, tracer);
};
