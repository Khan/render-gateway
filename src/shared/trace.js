// @flow
import type {Tracer} from "@google-cloud/trace-agent";
import type {Logger, ITraceSession, TraceSessionInfo} from "./types.js";

/**
 * Start tracing an event.
 *
 * This will log the start of a trace and open a trace session, which is
 * returned. Use the returned session to end the trace when the traced event is
 * over. The traced event will be logged and also written to the Google Cloud
 * StackDriver Trace agent.
 *
 * @param {Logger} logger A logger to use for documention and timing the
 * traced action.
 * @param {string} name The name of the traced action.
 * @param {Tracer} [tracer] A Google Cloud trace agent tracer which
 * can be used to record the traced action.
 * @returns {ITraceSession} A trace session that the caller should use to
 * indicate when the session is finished.
 */
export const trace = (
    logger: Logger,
    name: string,
    tracer?: Tracer,
): ITraceSession => {
    if (!name) {
        throw new Error("Must provide a name for the trace session.");
    }

    /**
     * We are going to use the logger's profiling API (provided by winston).
     * However, we want to mark the start of the trace as it gives us some
     * debug information which can be valuable when investigating operations.
     *
     * Winston only logs when profiling is done and the optional trace agent
     * tracer will only show the span if it is ended.
     *
     * Since this is noise in most situations, we will log this at the lowest
     * level of silly.
     */
    logger.silly(`TRACE: ${name}`);

    /**
     * Now we start the profiling timer.
     */
    const profiler = logger.startTimer();

    /**
     * Next, if we were given a tracer, we start a trace section for this so
     * trace session so that it will appear in Stackdriver Trace.
     *
     * We annotate the span with "TRACE:" so that it is clear in the trace
     * which spans were created by this API and which were inserted by other
     * means.
     */
    const span = tracer?.createChildSpan({name: `TRACE: ${name}`});

    /**
     * This is the function that we will return to our caller.
     * It can then be used to end and record the trace session.
     */
    const end = (info?: TraceSessionInfo): void => {
        /**
         * Let's mark our profile as done.
         *
         * We include the session info object, but make sure to set the level
         * and message ourselves.
         */
        profiler.done({
            ...info,
            message: `TRACED: ${name}`,
            level: info?.level || "debug",
        });

        /**
         * If we started a tracer span, let's end it.
         *
         * We disable this lint rule as the linter does not appear to
         * understand the optional chaining.
         */
        // eslint-disable-next-line flowtype/no-unused-expressions
        span?.endSpan();
    };

    return {
        get name() {
            return name;
        },
        end,
    };
};
