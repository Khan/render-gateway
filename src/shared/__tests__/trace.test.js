// @flow
import {trace} from "../trace.js";

describe("#trace", () => {
    it.each([[""], [null], [undefined]])(
        "should throw if the name is empty/falsy",
        (testPoint) => {
            // Arrange
            const fakeLogger = ({}: any);

            // Act
            const underTest = () => trace(fakeLogger, testPoint);

            // Assert
            expect(underTest).toThrowErrorMatchingSnapshot();
        },
    );

    it("should log the start of the session", () => {
        // Arrange
        const fakeLogger = ({
            silly: jest.fn(),
            startTimer: jest.fn(),
        }: any);

        // Act
        trace(fakeLogger, "SESSION_NAME");

        // Assert
        expect(fakeLogger.silly).toHaveBeenCalledWith("TRACE: SESSION_NAME");
    });

    it("should start a timer", () => {
        // Arrange
        const fakeLogger = ({
            silly: jest.fn(),
            startTimer: jest.fn(),
        }: any);

        // Act
        trace(fakeLogger, "SESSION_NAME");

        // Assert
        expect(fakeLogger.startTimer).toHaveBeenCalled();
    });

    it("should create a tracer span", () => {
        // Arrange
        const fakeLogger: any = {
            silly: jest.fn(),
            startTimer: jest.fn(),
        };
        const fakeTracer: any = {
            createChildSpan: jest.fn(),
        };

        // Act
        trace(fakeLogger, "SESSION_NAME", fakeTracer);

        // Assert
        expect(fakeTracer.createChildSpan).toHaveBeenCalledWith({
            name: "TRACE: SESSION_NAME",
        });
    });

    it("should return a trace session", () => {
        // Arrange
        const fakeLogger: any = {
            silly: jest.fn(),
            startTimer: jest.fn(),
        };

        // Act
        const result = trace(fakeLogger, "SESSION_NAME");

        // Assert
        expect(result).toBeDefined();
    });

    describe("trace session", () => {
        describe("#name", () => {
            it("should give name of the trace", () => {
                // Arrange
                const fakeLogger: any = {
                    silly: jest.fn(),
                    startTimer: jest.fn(),
                };
                const session = trace(fakeLogger, "SESSION_NAME");

                // Act
                const result = session.name;

                // Assert
                expect(result).toBe("SESSION_NAME");
            });

            it("should be read-only", () => {
                // Arrange
                const fakeLogger: any = {
                    silly: jest.fn(),
                    startTimer: jest.fn(),
                };
                const session = trace(fakeLogger, "SESSION_NAME");

                // Act
                /**
                 * Flow is going to complain because name is readonly.
                 * We will suppress this because we want to test that runtime
                 * code is enforcing this. The idea of this test is to ensure
                 * that folks don't edit out this specific behavior.
                 * $ExpectError
                 */
                const underTest = () => (session.name = "NEW_NAME!");

                // Assert
                expect(underTest).toThrowErrorMatchingInlineSnapshot(
                    `"Cannot set property name of #<Object> which has only a getter"`,
                );
            });
        });

        describe("#end", () => {
            it("should stop the profile timer", () => {
                // Arrange
                const fakeTimer = {
                    done: jest.fn(),
                };
                const fakeLogger: any = {
                    silly: jest.fn(),
                    startTimer: jest.fn().mockReturnValue(fakeTimer),
                };
                const session = trace(fakeLogger, "SESSION_NAME");

                // Act
                session.end();

                // Assert
                expect(fakeTimer.done).toHaveBeenCalledWith({
                    message: "TRACED: SESSION_NAME",
                    level: "debug",
                });
            });

            it("should use the log level given", () => {
                // Arrange
                const fakeTimer = {
                    done: jest.fn(),
                };
                const fakeLogger: any = {
                    silly: jest.fn(),
                    startTimer: jest.fn().mockReturnValue(fakeTimer),
                };
                const session = trace(fakeLogger, "SESSION_NAME");

                // Act
                session.end({level: "silly"});

                // Assert
                expect(fakeTimer.done).toHaveBeenCalledWith({
                    message: "TRACED: SESSION_NAME",
                    level: "silly",
                });
            });

            it("should use the given metadata info", () => {
                // Arrange
                const fakeTimer = {
                    done: jest.fn(),
                };
                const fakeLogger: any = {
                    silly: jest.fn(),
                    startTimer: jest.fn().mockReturnValue(fakeTimer),
                };
                const session = trace(fakeLogger, "SESSION_NAME");

                // Act
                session.end({
                    cached: true,
                    size: 56,
                    someotherrubbish: "blahblahblah",
                });

                // Assert
                expect(fakeTimer.done).toHaveBeenCalledWith({
                    message: "TRACED: SESSION_NAME",
                    level: "debug",
                    cached: true,
                    size: 56,
                    someotherrubbish: "blahblahblah",
                });
            });

            it("should not allow message to be overridden", () => {
                // Arrange
                const fakeTimer = {
                    done: jest.fn(),
                };
                const fakeLogger: any = {
                    silly: jest.fn(),
                    startTimer: jest.fn().mockReturnValue(fakeTimer),
                };
                const session = trace(fakeLogger, "SESSION_NAME");

                // Act
                session.end({
                    message: "I secretly want to change the message",
                });

                // Assert
                expect(fakeTimer.done).toHaveBeenCalledWith({
                    message: "TRACED: SESSION_NAME",
                    level: "debug",
                });
            });

            it("should end the trace span", () => {
                // Arrange
                const fakeTraceSpan = {
                    endSpan: jest.fn(),
                };
                const fakeTracer: any = {
                    createChildSpan: jest.fn().mockReturnValue(fakeTraceSpan),
                };
                const fakeTimer = {
                    done: jest.fn(),
                };
                const fakeLogger: any = {
                    silly: jest.fn(),
                    startTimer: jest.fn().mockReturnValue(fakeTimer),
                };
                const session = trace(fakeLogger, "SESSION_NAME", fakeTracer);

                // Act
                session.end();

                // Assert
                expect(fakeTraceSpan.endSpan).toHaveBeenCalled();
            });
        });
    });
});
