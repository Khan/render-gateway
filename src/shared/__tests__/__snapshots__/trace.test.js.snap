// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#trace should throw if the name is empty/falsy 1`] = `"Must provide a name for the trace session."`;

exports[`#trace should throw if the name is empty/falsy 2`] = `"Must provide a name for the trace session."`;

exports[`#trace should throw if the name is empty/falsy 3`] = `"Must provide a name for the trace session."`;
